### PR TITLE
feat(react-motions): add PresenceGroup

### LINKED ISSUE #30546: [Feature]: implement PresenceGroup in @fluentui/react-motions-preview
### Library

React Components / v9 (@fluentui/react-components)

### Describe the feature that you would like added

We need to support grouping presences similarly to `TransitionGroup` from `react-transition-group`. There is currently no implementation of this that uses WA API.

For context & API proposal, check #29958.

### Have you discussed this feature with our team

@microsoft/teams-prg

### Additional context

_No response_

### Validations

- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Priority

High

### PR DESCRIPTION
**Pull request by layershifter**
### `createPresenceComponent()` 

The factory was updated to store options that should not trigger re-renders in `optionsRef`.

### `PresenceGroup`

This PR adds a new component called `PresenceGroup`:

> `PresenceGroup` is a component that manages a set of components created with `createPresenceComponent()` in a list. `PresenceGroup` is a state machine for managing the mounting and unmounting of components over time.

While code is mostly ported from `react-transition-group`, it has some differences: we don't rely on cloning items and rely on React context instead.

Note: `PresenceGroup` is implemented as a class component to benefit from `getDerivedStateFromProps()`. While it's possible to implement this functionality with React hooks, it will require state management tricks.

## Related Issue(s)

Fixes #30546.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/802f530c89e5567ff0927a9669a987eb232e719e'>802f530c89e5567ff0927a9669a987eb232e719e</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-motions-preview/src/utils/groups/getNextChildMapping.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  Object.entries(childrenMapping).forEach(([key, childDefinition]) => {
+    const hasPrev = key in prevChildMapping;
+    const hasNext = key in nextChildMapping;
+
+    const prevChild = prevChildMapping[key];
+    const isLeaving = !prevChild?.visible ?? false;
+
+    // item is new (entering)
+    if (hasNext && (!hasPrev || isLeaving)) {
```
Comment (ling1726) [inline packages/react-components/react-motions-preview/src/utils/groups/getNextChildMapping.ts:18]: I don't understand the `hasNext && isLeaving` case, could you please clarify that one for me?

File: packages/react-components/react-motions-preview/src/utils/groups/mergeChildMappings.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    }
+
+    childMapping[nextKey] = getValueForKey(nextKey);
+  }
+
+  // Finally, add the keys which didn't appear before any key in `next`
+  for (let i = 0; i < pendingKeys.length; i++) {
+    childMapping[pendingKeys[i]] = getValueForKey(pendingKeys[i]);
+  }
```
Comment (ling1726) [inline packages/react-components/react-motions-preview/src/utils/groups/mergeChildMappings.ts:53]:
```suggestion
  for(const pendingKey of pendingKeys) {
    childMapping[pendingKey] = getValueForKey(pendingKey);
  }
```

File: packages/react-components/react-motions-preview/src/utils/groups/mergeChildMappings.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    return key in nextMapping ? nextMapping[key] : prevMapping[key];
+  }
+
+  // For each key of `next`, the list of keys to insert before that key in
+  // the combined list
+  const nextKeysPending: Record<string, string[]> = {};
+  let pendingKeys: string[] = [];
+
+  // eslint-disable-next-line guard-for-in
+  for (const prevKey in prevMapping) {
```
Comment (ling1726) [inline packages/react-components/react-motions-preview/src/utils/groups/mergeChildMappings.ts:22]: There's some assumptions made here about the order of children between states, I would add tests to make sure that it's not forgotten

File: packages/react-components/react-motions-preview/src/utils/groups/getNextChildMapping.ts (hunk 1 above)
Comment (layershifter) [reply packages/react-components/react-motions-preview/src/utils/groups/getNextChildMapping.ts]: I also, I added a few tests and restructured `if` conditions there.

**Review by ling1726 — APPROVED** [review]

**Pull request merged**